src/fund/components/FundButton.tsx

```diff
@@ -19,49 +31,122 @@ export function FundButton({
   rel,
   target,
   text: buttonText = 'Fund',
+  successText: buttonSuccessText = 'Success',
+  errorText: buttonErrorText = 'Something went wrong',
+  state: buttonState = 'default',
+  onPopupClose,
+  onClick,
 }: FundButtonReact) {
   const componentTheme = useTheme();
   // If the fundingUrl prop is undefined, fallback to our recommended funding URL based on the wallet type
-  const fundingUrlToRender = fundingUrl ?? useGetFundingUrl();
+  const fallbackFundingUrl = useGetFundingUrl();
+  const { address } = useAccount();
+  const fundingUrlToRender = fundingUrl ?? fallbackFundingUrl;
   const isDisabled = disabled || !fundingUrlToRender;
+  const shouldShowConnectWallet = !address;
+
+  const { startPopupMonitor } = usePopupMonitor(onPopupClose);
 
   const handleClick = useCallback(
     (e: React.MouseEvent) => {
       e.preventDefault();
+
       if (fundingUrlToRender) {
+        onClick?.();
         const { height, width } = getFundingPopupSize(
           popupSize,
           fundingUrlToRender,
         );
-        openPopup({
+        const popupWindow = openPopup({
           url: fundingUrlToRender,
           height,
           width,
           target,
         });
+
+        if (popupWindow) {
+          startPopupMonitor(popupWindow);
+        }
       }
     },
-    [fundingUrlToRender, popupSize, target],
+    [fundingUrlToRender, popupSize, target, onClick, startPopupMonitor],
   );
 
+  const buttonColorClass = useMemo(() => {
+    switch (buttonState) {
+      case 'error':
+        return background.error;
+      case 'loading':
+      case 'success':
+        return pressable.primary;
+      default:
+        return pressable.primary;
+    }
+  }, [buttonState]);
+
   const classNames = cn(
     componentTheme,
-    pressable.primary,
-    'px-4 py-3 inline-flex items-center justify-center space-x-2 disabled',
+    buttonColorClass,
+    'px-4 py-3 inline-flex items-center justify-center space-x-2',
     isDisabled && pressable.disabled,
     text.headline,
     border.radius,
     color.inverse,
     className,
   );
```

Comment:
the idiomatic way to write conditional values using the cn utility is with object syntax:
{
pressable.disabled: isDisabled,
}